packages/auto-form/src/FormBuilder.tsx

```diff
@@ -0,0 +1,185 @@
+  fields: IConfigurationProperty[];
+  renderField: (props: IRenderFieldProps) => any;
+  initialValues?: object;
+}
+
+export class FormBuilder extends React.Component<
+  IFormBuilderProps,
+  IFormBuilderState
+> {
+  public renderField = (props: IRenderFieldProps): any => {
+    // Massage the value in the definition to an input type
+    let type = props.property.type;
+    switch (type) {
+      case 'string':
+        type = 'text';
+        break;
+      case 'boolean':
+        type = 'checkbox';
+    }
+    if (props.property.enum && props.property.enum.length) {
+      type = 'select';
+    }
+    switch (type) {
+      case 'text':
+      case 'time':
+      case 'date':
+      case 'duration':
+      case 'color':
+      case 'datetime-local':
+      case 'email':
+      case 'file':
+      case 'month':
+      case 'range':
+      case 'tel':
+      case 'url':
+      case 'week':
+      case 'search':
+        return (
+          <Field
+            key={props.property.name}
+            name={props.property.name}
+            type={type}
+            {...props}
+            component={InputComponent}
+          />
+        );
+      case 'select':
+        return (
+          <Field
+            key={props.property.name}
+            name={props.property.name}
+            type={'select'}
+            {...props}
+            component={SelectComponent}
+          />
+        );
+      case 'textarea':
+        return (
+          <Field
+            key={props.property.name}
+            name={props.property.name}
+            type={'select'}
+            {...props}
+            component={TextAreaComponent}
+          />
+        );
+      default:
+        return (
+          <pre key={props.property.name}>
+            {props.property.name} = {props.value}
+          </pre>
+        );
+    }
+  };
+  public render() {
+    const initialValues = { ...this.props.initialValues };
+    const fields = [] as IConfigurationProperty[];
+    Object.keys(this.props.definition).forEach(key => {
+      fields.push({ name: key, ...this.props.definition[key] });
+      if (!(key in initialValues)) {
+        initialValues[key] = this.props.definition[key].defaultValue;
+      }
+    });
+    return this.props.children({
+      fields: fields.sort((a, b) => {
+        const aOrder = (a.order || 0) as number;
+        const bOrder = (b.order || 0) as number;
+        return aOrder - bOrder;
+      }),
+      initialValues,
+      onSave: this.props.onSave,
+      renderField: this.renderField,
+    });
+  }
+}
+
+// TODO move these to the UI package
+const InputComponent = ({
+  field,
+  form: { touched, errors },
+  ...props
+}: /*TODO*/ {
+  [name: string]: any;
+}) => (
+  <div className="form-group">
+    <label className="col-sm-2 control-label" htmlFor={field.name}>
+      {props.property.displayName}
+    </label>
+    <div className="col-sm-10">
+      <input data-testid={field.name} {...field} />
```

Comment:
Shouldn't `data-testid` be an `id` attribute? Otherwise there's no explicit link between labels and their respective form control

Reply:
+1, I just stuck `data-testid` in late-day to make at least one kinda working test.